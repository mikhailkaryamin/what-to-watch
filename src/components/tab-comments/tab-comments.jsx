import React, {
  PureComponent
} from 'react';
import {connect} from 'react-redux';
import {arrayOf} from 'prop-types';

import {commentPropTypes} from '../../types.js';

class TabComments extends PureComponent {
  constructor(props) {
    super(props);
  }

  render() {
    const {
      comments,
    } = this.props;

    return (
      <div className="movie-card__reviews movie-card__row">
        <div className="movie-card__reviews-col">

          {comments.filter((el, i) => (i + 1) % 2).map((comment) =>
            this._getMarkupComment(comment)
          )}

        </div>

        <div className="movie-card__reviews-col">

          {comments.filter((el, i) => !((i + 1) % 2)).map((comment) =>
            this._getMarkupComment(comment)
          )}

        </div>
      </div>
    );
  }

  _getDateComment(date) {
    const FORMAT_DATE = {
      year: `numeric`,
      day: `numeric`,
      month: `long`,
    };

    const dateComment = new Date(date);

    return dateComment.toLocaleString(`en-US`, FORMAT_DATE);
  }

  _getMarkupComment(comment) {
    return (
      <div
        key={`${comment.id}-review`}
        className="review"
      >
        <blockquote className="review__quote">
          <p className="review__text">
            {comment.text}
          </p>

          <footer className="review__details">
            <cite className="review__author">
              {comment.user.name}
            </cite>
            <time className="review__date" dateTime={new Date(comment.date)}>
              {this._getDateComment(comment.date)}
            </time>
          </footer>
        </blockquote>

        <div className="review__rating">
          {comment.rating}
        </div>
      </div>
    );
  }
}

TabComments.propTypes = {
  comments: arrayOf(commentPropTypes)
};

const mapStateToProps = (state) => ({
  comments: state.comments,
});

export {
  TabComments
};

export default connect(mapStateToProps)(TabComments);
