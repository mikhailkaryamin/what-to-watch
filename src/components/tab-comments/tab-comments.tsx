import * as React from 'react';
import {connect} from 'react-redux';

import {CommentType} from '../../types';
import {getComments} from '../../reducer/comment/selectors';

type Props = {
  comments: CommentType[],
}

const getDateCommentFormat = (date: number) => {
  const FORMAT_DATE = {
    year: `numeric`,
    day: `numeric`,
    month: `long`,
  };

  const dateComment = new Date(date);

  return dateComment.toLocaleString(`en-US`, FORMAT_DATE);
};

const getMarkupComment = (comment: CommentType) => {
  return (
    <div
      key={`${comment.id}-review`}
      className="review"
    >
      <blockquote className="review__quote">
        <p className="review__text">
          {comment.text}
        </p>

        <footer className="review__details">
          <cite className="review__author">
            {comment.user.name}
          </cite>
          <time className="review__date" dateTime={new Date(comment.date).toISOString()}>
            {getDateCommentFormat(comment.date)}
          </time>
        </footer>
      </blockquote>

      <div className="review__rating">
        {comment.rating}
      </div>
    </div>
  );
};

const TabComments: React.FC<Props> = (props: Props) => {
  const {
    comments,
  } = props;

  return (
    <div className="movie-card__reviews movie-card__row">
      <div className="movie-card__reviews-col">

        {comments.filter((el, i) => (i + 1) % 2).map((comment) =>
          getMarkupComment(comment)
        )}

      </div>

      <div className="movie-card__reviews-col">

        {comments.filter((el, i) => !((i + 1) % 2)).map((comment) =>
          getMarkupComment(comment)
        )}

      </div>
    </div>
  );
};

const mapStateToProps = (state: {}) => ({
  comments: getComments(state),
});

export {TabComments};

export default connect(mapStateToProps)(TabComments);
