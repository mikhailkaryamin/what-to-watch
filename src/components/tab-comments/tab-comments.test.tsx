import * as React from 'react';
import * as renderer from 'react-test-renderer';

import {TabComments} from './tab-comments';

describe(`Tab comments`, () => {
  const COMMENTS = [{
    id: 1,
    user: {
      id: 1,
      name: `Simple name`,
    },
    rating: 9,
    text: `some text`,
    date: 1293843434,
  },
  {
    id: 2,
    user: {
      id: 2,
      name: `Simple name`,
    },
    rating: 6,
    text: `some text`,
    date: 129384343,
  },
  {
    id: 3,
    user: {
      id: 3,
      name: `Simple name`,
    },
    rating: 9,
    text: `some text`,
    date: 12938034343183,
  }];

  const wrapper = renderer
    .create(
        <TabComments comments={COMMENTS}/>
    ).toJSON();

  test(`should render tab comments`, () => {
    expect(wrapper).toMatchSnapshot();
  });
});
