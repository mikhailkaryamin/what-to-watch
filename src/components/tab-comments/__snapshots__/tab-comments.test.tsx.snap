// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab comments should render tab comments 1`] = `
<div
  className="movie-card__reviews movie-card__row"
>
  <div
    className="movie-card__reviews-col"
  >
    <div
      className="review"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          some text
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Simple name
          </cite>
          <time
            className="review__date"
            dateTime="1970-01-15T23:24:03.434Z"
          >
            January 16, 1970
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        9
      </div>
    </div>
    <div
      className="review"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          some text
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Simple name
          </cite>
          <time
            className="review__date"
            dateTime="2379-12-28T18:25:43.183Z"
          >
            December 28, 2379
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        9
      </div>
    </div>
  </div>
  <div
    className="movie-card__reviews-col"
  >
    <div
      className="review"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          some text
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Simple name
          </cite>
          <time
            className="review__date"
            dateTime="1970-01-02T11:56:24.343Z"
          >
            January 2, 1970
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        6
      </div>
    </div>
  </div>
</div>
`;
